data/summary/types/summary.go

```diff
@@ -0,0 +1,284 @@
+	OutdatedSince     *time.Time `json:"outdatedSince" bson:"outdatedSince"`
+}
+
+type Bucket[T BucketData, S BucketDataPt[T]] struct {
+	Date           time.Time `json:"date" bson:"date"`
+	LastRecordTime time.Time `json:"lastRecordTime" bson:"lastRecordTime"`
+
+	Data S
+}
+
+type BucketDataPt[T BucketData] interface {
+	*T
+	CalculateStats(interface{}, *time.Time) error
+}
+
+func CreateBucket[T BucketData, A BucketDataPt[T]](t time.Time) *Bucket[T, A] {
+	bucket := new(Bucket[T, A])
+	bucket.Date = t
+	return bucket
+}
+
+type Buckets[T BucketData, S BucketDataPt[T]] []Bucket[T, S]
+
+type Stats interface {
+	CGMStats | BGMStats
+}
+
+type StatsPt[T Stats] interface {
+	*T
+	GetType() string
+	Init()
+	GetBucketsLen() int
+	GetBucketDate(int) time.Time
+	Update(any) error
+}
+
+type Summary[T Stats, A StatsPt[T]] struct {
+	ID     primitive.ObjectID `json:"-" bson:"_id,omitempty"`
+	Type   string
+	UserID string
+
+	Config Config `json:"config" bson:"config"`
+
+	Dates Dates `json:"dates" bson:"dates"`
+	Stats A
+}
+
+func NewConfig() Config {
+	return Config{
+		SchemaVersion:            1,
+		HighGlucoseThreshold:     highBloodGlucose,
+		VeryHighGlucoseThreshold: veryHighBloodGlucose,
+		LowGlucoseThreshold:      lowBloodGlucose,
+		VeryLowGlucoseThreshold:  veryLowBloodGlucose,
+	}
+}
+
+func (s Summary[T, A]) SetOutdated() {
+	s.Dates.OutdatedSince = pointer.FromTime(time.Now().UTC())
+}
+
+func NewDates() Dates {
+	return Dates{
+		HasLastUploadDate: false,
+		LastUploadDate:    time.Time{},
+		LastUpdatedDate:   time.Time{},
+		FirstData:         time.Time{},
+		LastData:          nil,
+		OutdatedSince:     nil,
+	}
+}
+
+func Create[T Stats, A StatsPt[T]](userId string) *Summary[T, A] {
+	s := new(Summary[T, A])
+	s.UserID = userId
+	s.Stats.Init()
+	s.Type = s.Stats.GetType()
+	s.Config = NewConfig()
+	s.Dates = NewDates()
+
+	return s
+}
+
+func GetTypeString[T Stats, A StatsPt[T]]() string {
+	s := new(Summary[T, A])
+	return s.Stats.GetType()
+}
+
+type Period interface {
+	BGMPeriod | CGMPeriod
+}
+
+func AddBin[T BucketData, A BucketDataPt[T], S Buckets[T, A]](buckets S, newStat Bucket[T, A]) error {
+	var hourCount int
+	var oldestHour time.Time
+	var oldestHourToKeep time.Time
+	var existingDay = false
+	var statsGap int
+	var newStatsTime time.Time
+
+	// update existing hour if one does exist
+	if len(buckets) > 0 {
+		for i := len(buckets) - 1; i >= 0; i-- {
+
+			if (buckets[i]).Date.Equal(newStat.Date) {
+				buckets[i] = newStat
+				existingDay = true
+				break
+			}
+
+			// we already passed our date, give up
+			if buckets[i].Date.After(newStat.Date) {
+				break
+			}
+		}
+
+		// add hours for any gaps that this new stat skipped
+		statsGap = int(newStat.Date.Sub(buckets[len(buckets)-1].Date).Hours())
+		for i := statsGap; i > 1; i-- {
+			newStatsTime = newStat.Date.Add(time.Duration(-i) * time.Hour)
+
+			buckets = append(buckets, *CreateBucket[T, A](newStatsTime))
```

Comment:
Don't think this will modify the length of the passed in buckets slice, but a copy of the slice. I believe you want either to pass in buckets as a pointer to a slice or return the new buckets slice and use that from any code that calls this function.